engineering/architecture/dateandtime.md

```diff
@@ -0,0 +1,107 @@
+
+## Overall input flow
+
+If the UI is able to parse a date, time, duration, etc., it should reformat it into the appropriate canonical form described below and send that as the input to be stored to the API. If it's not able to parse the input, it should send the raw input string to the API and let the service layer / database try to parse the string. This helps satisfy goal number 5. If the service layer is able to parse the string, then store it. Otherwise, raise a validation error.
+
+Another way to input data is through bulk import. In that case, we're completely reliant on the database parsing. This is the reason goal (4) is important.
+
+#### Regarding relative times
+
+There are some relative dates / times which make sense, e.g. "tomorrow", to store in a date column. These will _not_ be stored in a "relative" way, though. They'll be stored as a static date or time (one day later than the current day in the example). As for durations, there's some weird overlap between intervals "2 days ago" and a specific time "2 days ago". We'll choose whether to store the duration or a static time based on the column type.
+
+## Overall output flow
+
+For showing dates / times, they'll always be returned from the API in the appropriate canonical form described below. It's then up to the UI layer to format them for a pleasant display to the user.
+
+## RFC 3339 (based on ISO 8601)
+
+The canonical representation of a date, time, datetime, or duration in Mathesar will be as per the [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339#appendix-A) spec, with very minor modifications to match our use case.
+
+## Canonical Representation of Durations
+
+A duration describes the amount of time between two datetime values (or dates or times). Two durations can be added or subtracted to create another duration. A duration can be added to a date, time, or datetime to get a different date, time, or datetime. As per the RFC, the canonical representation of a duration is:
+
+```python
+f'P{years}Y{months}M{days}DT{hours}H{minutes}M{seconds}S'
+```
+Here, `seconds` can be an integer or float, and the other variables are all integers. For the canonical representation, we are filling in each unit with zeroes, which is optional in the RFC spec. To be completely proper, we aggregate units of time into larger ones where possible:
+
+- seconds aggregate into minutes,
+- minutes aggregate into hours, and
+- months aggregate into years.
+
+Because some units of time (months and days) are inconsistent w.r.t. the number of smaller units they contain, we avoid aggregating some units:
+
+- hours do not aggregate into days, and
+- days do not aggregate into months.
+
+#### Examples of Canonical Durations
+
+- `'1 year 1 month, 1 day 1 hour 1 minute 1.1 seconds'` maps to `P1Y1M1DT1H1M1.1S`
+- `'0 years 13 months 31 days 23 hours 60 minutes 61.1 seconds'` maps to `P1Y1M31DT24H1M1.1S`
```

Comment:
These are helpful examples of our expected input/output.

From reading these two examples, I infer:

- Commas do not matter
- Pluralization of words does not matter

Is this correct?

It might also be nice to strip the string `and` from the input, so that users can enter `1 year and 1 month`.

Reply:
Yes, that's correct (at least for the DB-layer parsing).  As for stripping "and", we might be able to do something for single inserts, but I don't think it would be very simple (with the default `INTERVAL` type) for `COPY`-based bulk imports.  We could probably add something as a feature at some point, but the implementation would be quite difficult.